In [1]:
# Initialize Otter
import otter
grader = otter.Notebook("hw8.ipynb")

# CPSC 330 - Applied Machine Learning

## Homework 8: Introduction to Computer vision, Time Series, and Survival Analysis (Lectures 19 to 20) 

**Due date: see the [Apr 07, 11:59 pm](https://github.com/UBC-CS/cpsc330-2024W2?tab=readme-ov-file#deliverable-due-dates-tentative).**

## Imports

In [2]:
from hashlib import sha1

import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

from sklearn.impute import SimpleImputer
from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline
from sklearn.preprocessing import StandardScaler, OrdinalEncoder, OneHotEncoder

from sklearn.linear_model import Ridge
from sklearn.ensemble import RandomForestRegressor

from sklearn.metrics import r2_score

<div class="alert alert-info">
    
## Submission instructions
<hr>
rubric={points:2}

Follow the [homework submission instructions](https://github.com/UBC-CS/cpsc330-2024W2/blob/main/docs/homework_instructions.md). 

**You may work in a group on this homework and submit your assignment as a group.** Below are some instructions on working as a group.  
- The maximum group size is 2. 
- Use group work as an opportunity to collaborate and learn new things from each other. 
- Be respectful to each other and make sure you understand all the concepts in the assignment well. 
- It's your responsibility to make sure that the assignment is submitted by one of the group members before the deadline. 
- You can find the instructions on how to do group submission on Gradescope [here](https://help.gradescope.com/article/m5qz2xsnjy-student-add-group-members).


When you are ready to submit your assignment do the following:

1. Run all cells in your notebook to make sure there are no errors by doing `Kernel -> Restart Kernel and Clear All Outputs` and then `Run -> Run All Cells`. 
2. Notebooks with cell execution numbers out of order will have marks deducted. Notebooks without the output displayed may not be graded at all (because we need to see the output in order to grade your work).
3. Upload the assignment using Gradescope's drag and drop tool. Check out this [Gradescope Student Guide](https://lthub.ubc.ca/guides/gradescope-student-guide/) if you need help with Gradescope submission.
4. Make sure that the plots and output are rendered properly in your submitted file. 
5. If the .ipynb file is too big and doesn't render on Gradescope, also upload a pdf or html in addition to the .ipynb.

<br><br>

## Exercise 1: time series prediction

In this exercise we'll be looking at a [dataset of avocado prices](https://www.kaggle.com/neuromusic/avocado-prices). You should start by downloading the dataset and storing it under the `data` folder. We will be forcasting average avocado price for the next week. 

In [3]:
df = pd.read_csv("data/avocado.csv", parse_dates=["Date"], index_col=0)
df.head()

Date  AveragePrice  Total Volume     4046       4225    4770  \
0 2015-12-27          1.33      64236.62  1036.74   54454.85   48.16   
1 2015-12-20          1.35      54876.98   674.28   44638.81   58.33   
2 2015-12-13          0.93     118220.22   794.70  109149.67  130.50   
3 2015-12-06          1.08      78992.15  1132.00   71976.41   72.58   
4 2015-11-29          1.28      51039.60   941.48   43838.39   75.78   

   Total Bags  Small Bags  Large Bags  XLarge Bags          type  year  region  
0     8696.87     8603.62       93.25          0.0  conventional  2015  Albany  
1     9505.56     9408.07       97.49          0.0  conventional  2015  Albany  
2     8145.35     8042.21      103.14          0.0  conventional  2015  Albany  
3     5811.16     5677.40      133.76          0.0  conventional  2015  Albany  
4     6183.95     5986.26      197.69          0.0  conventional  2015  Albany

In [4]:
df.shape

(18249, 13)

In [5]:
df["Date"].min()

Timestamp('2015-01-04 00:00:00')

In [6]:
df["Date"].max()

Timestamp('2018-03-25 00:00:00')

It looks like the data ranges from the start of 2015 to March 2018 (~2 years ago), for a total of 3.25 years or so. Let's split the data so that we have a 6 months of test data.

In [7]:
split_date = '20170925'
df_train = df[df["Date"] <= split_date]
df_test  = df[df["Date"] >  split_date]

In [8]:
assert len(df_train) + len(df_test) == len(df)

<br><br>

<!-- BEGIN QUESTION -->

### 1.1 How many time series? 
rubric={points:4}

In the [Rain in Australia](https://www.kaggle.com/datasets/jsphyg/weather-dataset-rattle-package) dataset from lecture demo, we had different measurements for each Location. 

We want you to consider this for the avocado prices dataset. For which categorical feature(s), if any, do we have separate measurements? Justify your answer by referencing the dataset.

<div class="alert alert-warning">

Solution_1.1
    
</div>

_Points:_ 4

type and region, we will have different time series for avocados from different region and different types of avocados. It does not make sense to split time series by any time related categories.

In [9]:
df["type"]

0     conventional
1     conventional
2     conventional
3     conventional
4     conventional
          ...     
7          organic
8          organic
9          organic
10         organic
11         organic
Name: type, Length: 18249, dtype: object

In [10]:
df["type"].nunique()

2

conventional or organic so this feature splits the time series into 2 sub time series

In [11]:
df["region"]

0               Albany
1               Albany
2               Albany
3               Albany
4               Albany
            ...       
7     WestTexNewMexico
8     WestTexNewMexico
9     WestTexNewMexico
10    WestTexNewMexico
11    WestTexNewMexico
Name: region, Length: 18249, dtype: object

In [12]:
df["region"].nunique()

54

54 different regions. This splits the time series into 54 sub time series.

<!-- END QUESTION -->

<br><br>

<!-- BEGIN QUESTION -->

### 1.2 Equally spaced measurements? 
rubric={points:4}

In the Rain in Australia dataset, the measurements were generally equally spaced but with some exceptions. How about with this dataset? Justify your answer by referencing the dataset.

<div class="alert alert-warning">

Solution_1.2
    
</div>

_Points:_ 4

Date is equally spaced with 7 days interval.

In [13]:
df_sorted = df.sort_values(by=["region", "type", "Date"])

In [14]:
alb_conv_df = df_sorted[(df_sorted["region"] == "Albany")  & (df_sorted["type"] == "conventional")]

In [15]:
alb_conv_df.value_counts()

Date        AveragePrice  Total Volume  4046      4225       4770     Total Bags  Small Bags  Large Bags  XLarge Bags  type          year  region
2015-01-04  1.22          40873.28      2819.50   28287.42   49.90    9716.46     9186.93     529.53      0.00         conventional  2015  Albany    1
2017-03-26  1.16          122951.65     3184.29   100854.22  70.17    18842.97    7252.70     11590.27    0.00         conventional  2017  Albany    1
2017-01-29  1.31          95424.59      3844.62   78315.15   484.56   12780.26    12393.84    382.06      4.36         conventional  2017  Albany    1
2017-02-05  1.49          183549.08     5666.60   165530.03  2119.02  10233.43    10047.32    186.11      0.00         conventional  2017  Albany    1
2017-02-12  1.42          97215.94      3808.43   81070.44   175.60   12161.47    11650.91    510.56      0.00         conventional  2017  Albany    1
                                                                                                   

In [16]:
alb_conv_df["Date"].diff()

51      NaT
50   7 days
49   7 days
48   7 days
47   7 days
      ...  
4    7 days
3    7 days
2    7 days
1    7 days
0    7 days
Name: Date, Length: 169, dtype: timedelta64[ns]

In [17]:
alb_conv_df["Date"].diff().value_counts()

Date
7 days    168
Name: count, dtype: int64

168 is one less than 169 but that is because there is one NaT at the beginning.

<!-- END QUESTION -->

<br><br>

<!-- BEGIN QUESTION -->

### 1.3 Interpreting regions 
rubric={points:4}

In the Rain in Australia dataset, each location was a different place in Australia. For this dataset, look at the names of the regions. Do you think the regions are also all distinct, or are there overlapping regions? Justify your answer by referencing the data.

<div class="alert alert-warning">

Solution_1.3
    
</div>

_Points:_ 4

In [18]:
unique_regions = df["region"].unique()

In [19]:
unique_regions

array(['Albany', 'Atlanta', 'BaltimoreWashington', 'Boise', 'Boston',
       'BuffaloRochester', 'California', 'Charlotte', 'Chicago',
       'CincinnatiDayton', 'Columbus', 'DallasFtWorth', 'Denver',
       'Detroit', 'GrandRapids', 'GreatLakes', 'HarrisburgScranton',
       'HartfordSpringfield', 'Houston', 'Indianapolis', 'Jacksonville',
       'LasVegas', 'LosAngeles', 'Louisville', 'MiamiFtLauderdale',
       'Midsouth', 'Nashville', 'NewOrleansMobile', 'NewYork',
       'Northeast', 'NorthernNewEngland', 'Orlando', 'Philadelphia',
       'PhoenixTucson', 'Pittsburgh', 'Plains', 'Portland',
       'RaleighGreensboro', 'RichmondNorfolk', 'Roanoke', 'Sacramento',
       'SanDiego', 'SanFrancisco', 'Seattle', 'SouthCarolina',
       'SouthCentral', 'Southeast', 'Spokane', 'StLouis', 'Syracuse',
       'Tampa', 'TotalUS', 'West', 'WestTexNewMexico'], dtype=object)

TotalUS and West looks like overlapping with other regions

<!-- END QUESTION -->

<br><br>

We will use the entire dataset despite any location-based weirdness uncovered in the previous part.

We will be trying to forecast the avocado price. The function below is adapted from [Lecture 19](https://github.com/UBC-CS/cpsc330-2024W2/tree/main/lectures), with some improvements.

In [20]:
def create_lag_feature(df, orig_feature, lag, groupby, new_feature_name=None, clip=False):
    """
    Creates a new feature that's a lagged version of an existing one.
    
    NOTE: assumes df is already sorted by the time columns and has unique indices.
    
    Parameters
    ----------
    df : pandas.core.frame.DataFrame
        The dataset.
    orig_feature : str
        The column name of the feature we're copying
    lag : int
        The lag; negative lag means values from the past, positive lag means values from the future
    groupby : list
        Column(s) to group by in case df contains multiple time series
    new_feature_name : str
        Override the default name of the newly created column
    clip : bool
        If True, remove rows with a NaN values for the new feature
    
    Returns
    -------
    pandas.core.frame.DataFrame
        A new dataframe with the additional column added.
        
    """
        
    if new_feature_name is None:
        if lag < 0:
            new_feature_name = "%s_lag%d" % (orig_feature, -lag)
        else:
            new_feature_name = "%s_ahead%d" % (orig_feature, lag)
    
    new_df = df.assign(**{new_feature_name : np.nan})
    for name, group in new_df.groupby(groupby):        
        if lag < 0: # take values from the past
            new_df.loc[group.index[-lag:],new_feature_name] = group.iloc[:lag][orig_feature].values
        else:       # take values from the future
            new_df.loc[group.index[:-lag], new_feature_name] = group.iloc[lag:][orig_feature].values
            
    if clip:
        new_df = new_df.dropna(subset=[new_feature_name])
        
    return new_df

We first sort our dataframe properly:

In [21]:
df_sort = df.sort_values(by=["region", "type", "Date"]).reset_index(drop=True)
df_sort

Date  AveragePrice  Total Volume     4046      4225    4770  \
0     2015-01-04          1.22      40873.28  2819.50  28287.42   49.90   
1     2015-01-11          1.24      41195.08  1002.85  31640.34  127.12   
2     2015-01-18          1.17      44511.28   914.14  31540.32  135.77   
3     2015-01-25          1.06      45147.50   941.38  33196.16  164.14   
4     2015-02-01          0.99      70873.60  1353.90  60017.20  179.32   
...          ...           ...           ...      ...       ...     ...   
18244 2018-02-25          1.57      18421.24  1974.26   2482.65    0.00   
18245 2018-03-04          1.54      17393.30  1832.24   1905.57    0.00   
18246 2018-03-11          1.56      22128.42  2162.67   3194.25    8.93   
18247 2018-03-18          1.56      15896.38  2055.35   1499.55    0.00   
18248 2018-03-25          1.62      15303.40  2325.30   2171.66    0.00   

       Total Bags  Small Bags  Large Bags  XLarge Bags          type  year  \
0         9716.46     9186.93      529.53          0.0  conventional  2015   
1         8424.77     8036.04      388.73          0.0  conventional  2015   
2        11921.05    11651.09      269.96          0.0  conventional  2015   
3        10845.82    10103.35      742.47          0.0  conventional  2015   
4         9323.18     9170.82      152.36          0.0  conventional  2015   
...           ...         ...         ...          ...           ...   ...   
18244    13964.33    13698.27      266.06          0.0       organic  2018   
18245    13655.49    13401.93      253.56          0.0       organic  2018   
18246    16762.57    16510.32      252.25          0.0       organic  2018   
18247    12341.48    12114.81      226.67          0.0       organic  2018   
18248    10806.44    10569.80      236.64          0.0       organic  2018   

                 region  
0                Albany  
1                Albany  
2                Albany  
3                Albany  
4                Albany  
...                 ...  
18244  WestTexNewMexico  
18245  WestTexNewMexico  
18246  WestTexNewMexico  
18247  WestTexNewMexico  
18248  WestTexNewMexico  

[18249 rows x 13 columns]

We then call `create_lag_feature`. This creates a new column in the dataset `AveragePriceNextWeek`, which is the following week's `AveragePrice`. We have set `clip=True` which means it will remove rows where the target would be missing.

In [22]:
df_hastarget = create_lag_feature(df_sort, "AveragePrice", +1, ["region", "type"], "AveragePriceNextWeek", clip=True)
df_hastarget

Date  AveragePrice  Total Volume     4046      4225    4770  \
0     2015-01-04          1.22      40873.28  2819.50  28287.42   49.90   
1     2015-01-11          1.24      41195.08  1002.85  31640.34  127.12   
2     2015-01-18          1.17      44511.28   914.14  31540.32  135.77   
3     2015-01-25          1.06      45147.50   941.38  33196.16  164.14   
4     2015-02-01          0.99      70873.60  1353.90  60017.20  179.32   
...          ...           ...           ...      ...       ...     ...   
18243 2018-02-18          1.56      17597.12  1892.05   1928.36    0.00   
18244 2018-02-25          1.57      18421.24  1974.26   2482.65    0.00   
18245 2018-03-04          1.54      17393.30  1832.24   1905.57    0.00   
18246 2018-03-11          1.56      22128.42  2162.67   3194.25    8.93   
18247 2018-03-18          1.56      15896.38  2055.35   1499.55    0.00   

       Total Bags  Small Bags  Large Bags  XLarge Bags          type  year  \
0         9716.46     9186.93      529.53          0.0  conventional  2015   
1         8424.77     8036.04      388.73          0.0  conventional  2015   
2        11921.05    11651.09      269.96          0.0  conventional  2015   
3        10845.82    10103.35      742.47          0.0  conventional  2015   
4         9323.18     9170.82      152.36          0.0  conventional  2015   
...           ...         ...         ...          ...           ...   ...   
18243    13776.71    13553.53      223.18          0.0       organic  2018   
18244    13964.33    13698.27      266.06          0.0       organic  2018   
18245    13655.49    13401.93      253.56          0.0       organic  2018   
18246    16762.57    16510.32      252.25          0.0       organic  2018   
18247    12341.48    12114.81      226.67          0.0       organic  2018   

                 region  AveragePriceNextWeek  
0                Albany                  1.24  
1                Albany                  1.17  
2                Albany                  1.06  
3                Albany                  0.99  
4                Albany                  0.99  
...                 ...                   ...  
18243  WestTexNewMexico                  1.57  
18244  WestTexNewMexico                  1.54  
18245  WestTexNewMexico                  1.56  
18246  WestTexNewMexico                  1.56  
18247  WestTexNewMexico                  1.62  

[18141 rows x 14 columns]

Our goal is to predict `AveragePriceNextWeek`. 

Let's split the data:

In [23]:
df_train = df_hastarget[df_hastarget["Date"] <= split_date]
df_test  = df_hastarget[df_hastarget["Date"] >  split_date]

<br><br>

<!-- BEGIN QUESTION -->

### 1.4 `AveragePrice` baseline 
rubric={points}

Soon we will want to build some models to forecast the average avocado price a week in advance. Before we start with any ML though, let's try a baseline. Previously we used `DummyClassifier` or `DummyRegressor` as a baseline. This time, we'll do something else as a baseline: we'll assume the price stays the same from this week to next week. So, we'll set our prediction of "AveragePriceNextWeek" exactly equal to "AveragePrice", assuming no change. That is kind of like saying, "If it's raining today then I'm guessing it will be raining tomorrow". This simplistic approach will not get a great score but it's a good starting point for reference. If our model does worse that this, it must not be very good. 

Using this baseline approach, what $R^2$ do you get on the train and test data?

<div class="alert alert-warning">

Solution_1.4
    
</div>

_Points:_ 4

_Type your answer here, replacing this text._

In [24]:
#df_train 
#df_test

In [25]:
from sklearn.metrics import r2_score

In [26]:
train_r2 = r2_score(df_train["AveragePriceNextWeek"], df_train["AveragePrice"])

In [27]:
test_r2 = test_r2 = r2_score(df_test["AveragePriceNextWeek"], df_test["AveragePrice"])

In [28]:
train_r2

0.8285800937261841

In [29]:
test_r2

0.7631780188583048

In [30]:
assert not train_r2 is None, "Are you using the correct variable name?"
assert not test_r2 is None, "Are you using the correct variable name?"
assert sha1(str(round(train_r2, 3)).encode('utf8')).hexdigest() == 'b1136fe2a8918904393ab6f40bfb3f38eac5fc39', "Your training score is not correct. Are you using the right features?"
assert sha1(str(round(test_r2, 3)).encode('utf8')).hexdigest() == 'cc24d9a9b567b491a56b42f7adc582f2eefa5907', "Your test score is not correct. Are you using the right features?"

<!-- END QUESTION -->

<br><br>

<!-- BEGIN QUESTION -->

### 1.5 Forecasting average avocado price
rubric={points:10}

Now that the baseline is done, let's build some models to forecast the average avocado price a week later. Experiment with a few approachs for encoding the date. Justify the decisions you make. Which approach worked best? Report your test score and briefly discuss your results.

Benchmark: you should be able to achieve $R^2$ of at least 0.79 on the test set. I got to 0.80, but not beyond that. Let me know if you do better!

Note: because we only have 2 splits here, we need to be a bit wary of overfitting on the test set. Try not to test on it a ridiculous number of times. If you are interested in some proper ways of dealing with this, see for example sklearn's [TimeSeriesSplit](https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.TimeSeriesSplit.html), which is like cross-validation for time series data.

<div class="alert alert-warning">

Solution_1.5
    
</div>

_Points:_ 10

At some point we need to encode string features into numeric ones to fit RandomForest model on the data.
However just make sure that encoding happens after crate_lag_feature because create_lag_feature is relying on "type" and "region" columns for grouping.

In [31]:
df_lagged = create_lag_feature(df_hastarget, "AveragePrice", -1, ["region", "type"], "AveragePriceLag_1weekago")
df_lagged = create_lag_feature(df_lagged, "AveragePrice", -2, ["region", "type"], "AveragePriceLag_2weekago")
df_lagged = create_lag_feature(df_lagged, "AveragePrice", -3, ["region", "type"], "AveragePriceLag_3weekago")
#df_lagged = create_lag_feature(df_lagged, "AveragePrice", +1, ["region", "type"], "AveragePriceNextWeek", clip=True)

df_train = df_lagged[df_lagged["Date"] <= split_date].copy()
df_test = df_lagged[df_lagged["Date"] > split_date].copy()

In [32]:
df_train.loc[:,"year"] = df_train["Date"].dt.year
df_train.loc[:,"month"] = df_train["Date"].dt.month
df_train.loc[:,"day"] = df_train["Date"].dt.day
#df_train.loc[:,"dayofweek"] = df_train["Date"].dt.dayofweek


df_test.loc[:,"year"] = df_test["Date"].dt.year
df_test.loc[:,"month"] = df_test["Date"].dt.month
df_test.loc[:,"day"] = df_test["Date"].dt.day
#df_test.loc[:,"dayofweek"] = df_test["Date"].dt.dayofweek

captures seasonality, trend, cyclical patterns and so on

In [33]:
df_train = df_train.drop(columns=["Date"])

In [34]:
df_train_X = df_train.drop(columns=["AveragePriceNextWeek"])
df_train_Y = df_train["AveragePriceNextWeek"]
df_test_X = df_test.drop(columns=["AveragePriceNextWeek"])
df_test_Y =  df_test["AveragePriceNextWeek"]

In [35]:
numeric_features = ["AveragePrice", "Total Volume", "4046", "4225", "4770", "Total Bags", "Small Bags", "Large Bags", "XLarge Bags", "year", 
                     "AveragePriceLag_1weekago", "AveragePriceLag_2weekago", "AveragePriceLag_3weekago", "month","day"]
categorical_features = ["type", "region"]

In [36]:
df_train

AveragePrice  Total Volume     4046      4225    4770  Total Bags  \
0              1.22      40873.28  2819.50  28287.42   49.90     9716.46   
1              1.24      41195.08  1002.85  31640.34  127.12     8424.77   
2              1.17      44511.28   914.14  31540.32  135.77    11921.05   
3              1.06      45147.50   941.38  33196.16  164.14    10845.82   
4              0.99      70873.60  1353.90  60017.20  179.32     9323.18   
...             ...           ...      ...       ...     ...         ...   
18218          2.50      16137.93  2616.96   3672.96    0.00     9848.01   
18219          2.39       7657.47   927.27   4056.73    0.00     2673.47   
18220          2.38      11857.31  1562.10   4565.41    0.00     5729.80   
18221          2.36      10464.29  1845.14   2819.17    0.00     5799.98   
18222          2.26       9528.64  1545.34   2234.23    0.00     5749.07   

       Small Bags  Large Bags  XLarge Bags          type  year  \
0         9186.93      529.53          0.0  conventional  2015   
1         8036.04      388.73          0.0  conventional  2015   
2        11651.09      269.96          0.0  conventional  2015   
3        10103.35      742.47          0.0  conventional  2015   
4         9170.82      152.36          0.0  conventional  2015   
...           ...         ...          ...           ...   ...   
18218     9816.58       31.43          0.0       organic  2017   
18219     2629.18       44.29          0.0       organic  2017   
18220     5719.96        9.84          0.0       organic  2017   
18221     5796.65        3.33          0.0       organic  2017   
18222     5722.40       26.67          0.0       organic  2017   

                 region  AveragePriceNextWeek  AveragePriceLag_1weekago  \
0                Albany                  1.24                       NaN   
1                Albany                  1.17                      1.22   
2                Albany                  1.06                      1.24   
3                Albany                  0.99                      1.17   
4                Albany                  0.99                      1.06   
...                 ...                   ...                       ...   
18218  WestTexNewMexico                  2.39                      2.43   
18219  WestTexNewMexico                  2.38                      2.50   
18220  WestTexNewMexico                  2.36                      2.39   
18221  WestTexNewMexico                  2.26                      2.38   
18222  WestTexNewMexico                  2.37                      2.36   

       AveragePriceLag_2weekago  AveragePriceLag_3weekago  month  day  
0                           NaN                       NaN      1    4  
1                           NaN                       NaN      1   11  
2                          1.22                       NaN      1   18  
3                          1.24                      1.22      1   25  
4                          1.17                      1.24      2    1  
...                         ...                       ...    ...  ...  
18218                      2.01                      1.78      8   27  
18219                      2.43                      2.01      9    3  
18220                      2.50                      2.43      9   10  
18221                      2.39                      2.50      9   17  
18222                      2.38                      2.39      9   24  

[15441 rows x 18 columns]

In [37]:
from sklearn.compose import ColumnTransformer
from sklearn.preprocessing import StandardScaler, OneHotEncoder
from sklearn.pipeline import make_pipeline

In [38]:
preprocessor = ColumnTransformer(
    transformers=[
        ("standardscaler", make_pipeline(SimpleImputer(strategy="mean"), StandardScaler()), numeric_features),
        ("onehotencoder",OneHotEncoder(handle_unknown="ignore"), categorical_features)
    ]
)

In [39]:
rf_pipe = make_pipeline(
    (preprocessor),
    (RandomForestRegressor(n_estimators=20, max_depth=7,random_state=0))
)            

In [40]:
rf_pipe.fit(df_train_X, df_train_Y)

Pipeline(steps=[('columntransformer',
                 ColumnTransformer(transformers=[('standardscaler',
                                                  Pipeline(steps=[('simpleimputer',
                                                                   SimpleImputer()),
                                                                  ('standardscaler',
                                                                   StandardScaler())]),
                                                  ['AveragePrice',
                                                   'Total Volume', '4046',
                                                   '4225', '4770', 'Total Bags',
                                                   'Small Bags', 'Large Bags',
                                                   'XLarge Bags', 'year',
                                                   'AveragePriceLag_1weekago',
                                                   'AveragePriceLag_2weekago',
                                                   'AveragePriceLag_3weekago',
                                                   'month', 'day']),
                                                 ('onehotencoder',
                                                  OneHotEncoder(handle_unknown='ignore'),
                                                  ['type', 'region'])])),
                ('randomforestregressor',
                 RandomForestRegressor(max_depth=7, n_estimators=20,
                                       random_state=0))])

In [41]:
train_r2 = rf_pipe.score(df_train_X, df_train_Y)
test_r2 = rf_pipe.score(df_test_X, df_test_Y)

In [42]:
train_r2

0.8814663735969666

In [43]:
test_r2

0.7964680987951787

In [44]:
...

Ellipsis

In [45]:
...

Ellipsis

In [46]:
...

Ellipsis

In [47]:
...

Ellipsis

<!-- END QUESTION -->

<br><br><br><br>

## Exercise 2: Short answer questions

<!-- BEGIN QUESTION -->

### 2.1 Time series

rubric={points:6}

The following questions pertain to Lecture 20 on time series data:

1. Sometimes a time series has missing time points or, worse, time points that are unequally spaced in general. Give an example of a real world situation where the time series data would have unequally spaced time points.
2. In class we discussed two approaches to using temporal information: encoding the date as one or more features, and creating lagged versions of features. Which of these (one/other/both/neither) two approaches would struggle with unequally spaced time points? Briefly justify your answer.
3. When studying time series modeling, we explored several ways to encode date information as a feature for the citibike dataset. When we used time of day as a numeric feature, the Ridge model was not able to capture the periodic pattern. Why? How did we tackle this problem? Briefly explain.

<div class="alert alert-warning">

Solution_2.1
    
</div>

_Points:_ 6

1. The occurence of earthquake has unpredicable intervals.
2. Creating Lagged feature would be hard if the data is taken with unevenly spaced intervals because the feature creation is relying on the assumption that the data is laid out in a equally spaced intervals. For example, just as I have done for this assignment I extracted and created the lagged features with a week lag between each features, but it would be hard(or impossible) to get such intervals if the data samples are taken with uneven intervals.
3. Ridge regression model expects linearlity in the data samples so for example if the days of a week is expressed as 0~6 the ridge regression would falsely assume that 6 has more impact on our prediction, which is not correct. Ridge regression model basically does not capture the cyclical data.

<!-- END QUESTION -->

<br><br>

<!-- BEGIN QUESTION -->

### 2.2 Computer vision 
rubric={points:6}

The following questions pertain to Lecture 19 on multiclass classification and introduction to computer vision. 

1. How many parameters (coefficients and intercepts) will `sklearn`’s `LogisticRegression()` model learn for a four-class classification problem, assuming that you have 10 features? Briefly explain your answer.
2. In Lecture 19, we briefly discussed how neural networks are sort of like `sklearn`'s pipelines, in the sense that they involve multiple sequential transformations of the data, finally resulting in the prediction. Why was this property useful when it came to transfer learning?
3. Imagine that you have a small dataset with ~1000 images containing pictures and names of 50 different Computer Science faculty members from UBC. Your goal is to develop a reasonably accurate multi-class classification model for this task. Describe which model/technique you would use and briefly justify your choice in one to three sentences.

<div class="alert alert-warning">

Solution_2.2
    
</div>

_Points:_ 6

1. 4 classes -> 4 different regression model with 10 coefficients and one intercept (11). so 44 in totoal.
2. Layer structured neural network pre-trained model allows us to reuse previous layers. We take earlier layers to make our specific prediction because earlier layers are generic and can be applied to wide range of tasks.
3. I would use CNN models like VGG, fine tuning the last layer but keep the earlier layers. I would not train NN models from scratch because it normally overfits because of the lack of data, which is the case in this problem because we only have 1000 images

<!-- END QUESTION -->

<br><br>

<!-- BEGIN QUESTION -->

## 2.3 Survival analysis
<hr>

rubric={points:6}

The following questions pertain to Lecture 21 on survival analysis. We'll consider the use case of customer churn analysis.

1. What is the problem with simply labeling customers are "churned" or "not churned" and using standard supervised learning techniques?
2. Consider customer A who just joined last week vs. customer B who has been with the service for a year. Who do you expect will leave the service first: probably customer A, probably customer B, or we don't have enough information to answer? Briefly explain your answer. 
3. If a customer's survival function is almost flat during a certain period, how do we interpret that?

<div class="alert alert-warning">

Solution_2.3
    
</div>

_Points:_ 6

1. This approach does not take into consideration the time when the customers might churn, while survival analysis models the time until churn occurs and handle cencored data appropriately.
2. Probably customer A. but we can't tell exactly because hazard function takes into consideration more than just the time.
3. Hazard rate is very low. so the customer is less likey to churn.

<!-- END QUESTION -->

<br><br>

**Before submitting your assignment, please make sure you have followed all the instructions in the Submission instructions section at the top.** 

![](img/eva-well-done.png)